### ISSUE TITLE
Receiving a notification does not set the message info

### ISSUE THREAD

**Issue by philipderrin** [seL4/seL4#114]
When a message is received from a notification object, the message info register is left uninitialised. This has been the case since `doAsyncTransfer` was removed by commit 97042a0f0.

This is reasonable when you know that you're receiving from a notification object, since the message info is never useful in that case, and `seL4_Wait` doesn't return it (though `seL4_Poll` does?). So presumably this was not an accident and was done to save a couple of instructions in the notification path.

However, this also applies to delivery of a notification to a bound thread blocked in `seL4_Recv`, which would otherwise return with valid message info.

Therefore, a thread with a bound notification must check for the reserved badge values that indicate a notification _before_ it interprets the message info. This is often not a problem, but it makes writing a high-level language stub for `seL4_Recv` a bit awkward, as it's not safe to simply return the received MRs — there must be a separate binding for `seL4_GetMR`.

```ocaml
module Syscalls : sig
    type cptr
    type label
    type badge

    (* This can't be safely implemented *)
    val recv : cptr -> label * badge * (nativeint list)
end
```

Since the message info register is usually adjacent to the badge register, I think this should be very cheap to fix on any architecture with a store-pair instruction — maybe even free on AArch64, where a `str badge, [context, offset]` might become `stp badge, xzr, [context, offset]`.

**Comment by pingerino** [comment]
I can look into this, but unfortunately the change will require proof adjustments so it'll need to wait until the verification team are less overloaded. There is no fastpath for bound endpoints, so I doubt there will be any performance impact at all.

**Comment by canarysnort01** [comment]
I've been discussing this issue with the Robigalia team and we have some thoughts and additional solutions to consider.

### Existing situation and working around it

The first thing I thought when reading this was that the existing behavior might turn out to be a good thing. Before the syscall you could seed the message info register with a value that is impossible for the kernel to return when receiving an IPC. If the message info register still contains the invalid value after the syscall then you know for certain that you received a bound notification; otherwise, you received an IPC.

There are a couple ways you could construct an invalid message info register value:

1. Set the caps unwrapped field to 0 and the caps received field to a value greater than 1. This is invalid because seL4 only supports receiving at most one invokable, non-unwrapped capability.

2. Set the message length to a value greater than the maximum message length supported by the kernel. This is possible because the number of bits allocated to the message length field in the message info bitfield can hold a larger number than the maximum supported message length. 

This workaround is actually better than the official guidance to partition the badge bits into notification-only bits and endpoint-only bits in order to tell which was received because it allows you to use the full number of badge bits for both endpoint badges and bound notification flags.

This is fine for the Receive syscall but things get tricky for ReplyRecv and friends since the message info register must be used for sending the reply. 

I believe you could modify the ReplyRecv message info register in the following way in order to construct a message info register value that would be invalid to receive while still allowing the reply message to be sent:

1. Change the caps unwrapped field so that it indicates one unwrapped cap. This field is not meaningful during the send phase so setting it does not change the reply behavior.

2. Increase the number of caps sent to 3 and insert a CPtr to a null capability into the sent caps array for each of the additional caps. This will have no observable effect on the send operation because seL4 will silently ignore the null capabilities.

If a bound notification is received during the receive phase the message info register will sill contain the message info for the reply, which indicates 3 capabilities received but only 1 capability unwrapped, an impossible situation.

While I believe this would work, it would impose a performance penalty because the IPC buffer would be touched on every message and the fast path through the kernel would never be used.

### Additional Solutions

Solving this issue is important to prevent getting bit by Hyrum's Law[1]. If people start to rely on poorly defined register-level behavior to determine if a bound notification was received it could make it more difficult for you to change the meaning of message info in the future, such as increasing the number of invokable capabilities that can be received.

#### Explicitly define a message info sentinel value meaning "received notification"

You could explicitly set the message info register to a well-defined invalid value as I described above in the kernel when a bound notification is received. You would define this value as a constant in libsel4 and document its meaning in the manual.

This would not break existing customers because they are not currently using the message info register in this case.

Going forward, customers would be able to definitively tell whether a notification or an IPC was received and they could use the full badge space for both endpoint capability badges and bound notification signals.

Since you are defining the invalid value yourself for each kernel release as a constant in libsel4, you can change it as you change the meaning of message info.

One concern for this solution is that at least one invalid encoding of message info must be available. I feel this concern is mitigated by public statements that have been made regarding your desire to further constrain and simplify IPC functionality[2], which would likely increase the availability of unused bits in the message info encoding.

#### Even more explicit behavior

You could change the behavior of receiving a bound notification to the following:

- The badge register is filled with the badge value that is stored in the _capability to the notification that was orignially bound to the TCB_
- Message register 1 is filled with the notification word

Before binding a notification to their TCB, the customer would first badge it with the sentinel value they wish to use to indicate "received a bound notification".

Then, every time the customer calls receive they check to see if the badge value is equal to the sentinel value they picked. If it is, then they know they received a notification, and will read message register 1 to get the notification word. Otherwise, they know they received an IPC.

This only requires the customer to reserve a single value from the endpoint capability badge space to mean "received a bound notification". The full number of badge bits can be used to signal the bound notification.

This solution would be a breaking API change.

### Conclusion

We would prefer a solution that would allow us to use to use the full number of badge bits for signaling bound notifications.

### References

[1] https://www.hyrumslaw.com
[2] https://microkerneldude.wordpress.com/2019/03/07/how-to-and-how-not-to-use-sel4-ipc/

**Comment by ratmice** [comment]
Say we have a threads `A` and `B` where A is communicating asynchronously with B, and B both sends notifications to A as well as  blocks sending IPC.  Isn't it the case that when `A` Polls, it can receive multiple notifications + messages, in cases where this is possible it seems like you must partition the endpoint/badge bits.  If we expand this to `A1, A2` communicating to `B` on the same endpoint notifications from A2 could arrive after sends from A1.

I guess I just don't see how distinguishing these via  message info can work in these cases, and doesn't end up complicating such scenerios.

**Comment by canarysnort01** [comment]
If A does a receive and there exists both an active bound notification and an IPC message waiting to be received, A will only receive the bound notification word; the IPC message will not be delivered until A calls receive again.

**Comment by canarysnort01** [comment]
The message info would only indicate whether a bound notification was received or the IPC message was received; the badge is still available to indicate the meaning, and you'd have the extra information to know whether it's an endpoint badge or a notifcation word, allowing you to use the full badge bits for both instances.

**Comment by ratmice** [comment]
Thanks for clarifying that aspect, that is the detail I was missing.

**Comment by canarysnort01** [comment]
I updated the ReplyRecv workaround with a simpler one.

If the number of capabilities sent is less than the maximum, set the caps unwrapped bit for a capability greater than the number that was sent.

If the number of capabilities sent is the maximum, set the caps unwrapped field to 0.

This avoids the performance implications of touching the IPC buffer and allows the kernel fast path to be used.

**Comment by canarysnort01** [comment]
So, I didn't think through the existing situation enough. I had thought that without the kernel telling us whether we received a bound notification or IPC message that we would have to partition the badge space into notification-only bits and endpoint-only bits.

But, thinking through it some more, I realized we only really need to reserve a single bit in the badge space for this indicator. For example, we could have a convention where we always set the highest badge bit on a notification but never on an endpoint.

This would reduce our total badge space by half but allow us to use all of the remaining bits for both notification flags and endpoint capability badges.

Even so, I still think it would be preferable if the kernel was explicit about whether it received a bound notification or an IPC message.

**Comment by pingerino** [comment]
Once upon a time we used the message label to indicate what type of message came through (notification or IPC) using the message label. We removed this in favour of letting the user decide the policy, also because the label can be easily forged (@AdrianDanis can you confirm?). Generally the pattern is setting 1 bit on the top badge indicate the type of the  object - notification or endpoint. That leaves the rest of the badge bits for ids.

**Comment by canarysnort01** [comment]
I see.

I believe the kernel should communicate this in a way that cannot be forged. I think the additional solutions I brought up would achieve this without introducing policy into the kernel. 

The first solution would introduce no policy into the kernel nor require the user to set a policy.

The second solution would introduce no policy into the kernel and would only require the user to have a policy on a single value in the endpoint badge space to represent a received bound notification while allowing all the bits to be used to signal the bound notification.

These are much cheaper than requiring the user to sacrifice half of the badge space to receive bound notifications.

**Comment by pingerino** [comment]
I'm curious: what's the scenario where you need more than 2^27 unique badges (or 2^63 on 64 bit?)
 
The awkwardness of the bound notification API has been brought up many times, but ultimately the effort to impact ratio of updating the proofs and reworking the models for something that can easily be worked around at user-level has always come out as a non-starter due to resource constraints.

**Comment by kent-mcleod** [comment]
@canarysnort01 there is a Jira issue for whether recv operations should be able to distinguish between endpoint and notification messages: https://sel4.atlassian.net/browse/SELFOUR-189.  Currently it is expected that a convention is used that reserves a bit of the badge space to distinguish between Endpoints and Notifications.  There is a separate Jira issue for considering the number of badge bits being a limitation for notifications: https://sel4.atlassian.net/browse/SELFOUR-190 (Both of these issues are old and are from before Notifications were renamed from async-endpoints).  Neither issue has becoming a limiting design constraint yet so they haven't been prioritized so far.

**Comment by canarysnort01** [comment]
@kent-mcleod Thanks for those links. I'll take a look at them and see if I can add anything.

@pingerino Our concern is mostly around the lack of an unforgable indicator from the kernel on whether we received an IPC message or notification.

We want to build dynamic systems on top of seL4.

Like the OP, we don't want to rely on a convention (which bits should mean what) to be able to safely decode a message, especially if we're going to be passing around endpoint caps at runtime between arbitrary code.

I don't believe setting the message register to 0 goes far enough to solve the problem. In fact, it may even be worse for us since I may have found a roundabout way to use the register to infer an unforgable indictator from the kernel, although I'd prefer an officially supported way to get this.

**Comment by kent-mcleod** [comment]
One of the downsides of having the kernel provide unforgable information is that it may prevent you from being able to do cap virtualisation.

Assume you have a thread A that is being intermediated (by a security monitor or some other virtualisation layer). The intermediary B intercepts A’s system calls and executes them on A’s behalf. This scheme is easy to implement in a cap system: all A’s caps are in fact endpoints to the intermediary, which contains the real objects. In this design, a notification object may also be virtualized as an endpoint and so the security monitor would need to be able to send a message that is indistinguishable from a notification.

I'm not sure whether supporting cap virtualization is still a design goal though.  A couple years ago @gernotheiser had a student that researched designs involving this but I'm not sure what the results were.

**Comment by gernotheiser** [comment]
Very good point @kent-mcleod. I agree, we should not do anything that prevents caps from being virtualised. We should rather eliminate the features that break virtualisation.

**Comment by canarysnort01** [comment]
@kent-mcleod I understand the need to do this when a client is waiting on what it thinks is a notification, but is actually an endpoint to an interposing server. In that scenario the interposing server would need to be able to send to that endpoint and have its message be interpreted as a notification. I am OK with that, because the client is only expecting to receive a notification in that context.

The only time I'd like the kernel to explicitly say that the message is a notification is if the client is waiting on an endpoint but the kernel is delivering a bound notification.

It seem to me that if an interposing server wants to deliver a bound notification to a client, it would need to actually bind a notification to the client's thread and then signal it, which would get delivered by the kernel.

**Comment by canarysnort01** [comment]
@kent-mcleod To support virtualization you might need to start delivering bound notifications when someone is waiting on a notification, in order to make the possible things that could happen when receiving against an endpoint and a notification the same so, but having an unforgable kernel indicator of a bound notification receipt would still be virtualizable and even more valuable in this situation.

**Comment by kent-mcleod** [comment]
> @kent-mcleod To support virtualization you might need to start delivering bound notifications when someone is waiting on a notification, in order to make the possible things that could happen when receiving against an endpoint and a notification the same so, but having an unforgable kernel indicator of a bound notification receipt would still be virtualizable and even more valuable in this situation.

The unforgeable indicator would prevent the emulation of notification delivery to be performed by sending on an endpoint. 

> @pingerino Our concern is mostly around the lack of an unforgable indicator from the kernel on whether we received an IPC message or notification.

It might be easier to discuss trade-offs if you are able to share a specific example where this is a limitation?  The purpose of badges is to give the receiver a way to authenticate the senders.  This relies on the receiver being able to trust that a malicious sender can't change the badge that it is sending on so that it can trust the badges.  My understanding is that you have a scenario where the receiver cannot trust the badge value and requires the kernel to provide an extra unforgeable bit?

**Comment by canarysnort01** [comment]
> The unforgeable indicator would prevent the emulation of notification delivery to be performed by sending on an endpoint.

No, it would not.

Again, the unforgable indicator is only to indicate that you were signaled by the TCB-bound notification _instead of_ receiving the IPC message that you were expecting, it is not for notification receipt in general.

> It might be easier to discuss trade-offs if you are able to share a specific example where this is a limitation?

This is a limitation in that without the indicator there is no way to know when receiving on an Endpoint whether you received an IPC Message or a bound notification without interpreting the badge value.

We are building dynamic systems on seL4 and so each application might want to use its own convention but still share objects. For example, some code might want to send an endpoint to some other code with the intent that they receive on it.

The issue is decoding an IPC message becomes unsafe or error-prone without knowing the convention in use; but, in a dynamic system each application may wish to use their own convention, and this becomes a problem when they share capabilites.

**Comment by kent-mcleod** [comment]
> No, it would not.

I meant that without the change, it would be possible to cap-virtualize a component by only using an endpoint and no notification object.  Having the forgeable indicator would make this impossible. I don't have an opinion on whether needing to do that is a good design or not; just trying to identify it as a consequence.

> We are building dynamic systems on seL4 and so each application might want to use its own convention but still share objects. For example, some code might want to send an endpoint to some other code with the intent that they receive on it.

So if you were to receive on an endpoint that you didn't control the badge allocation for, but also had a bound notification that you did control the badge allocation for, you couldn't trust that a message you were receiving was from the notification or endpoint.
Are you able to be more specific? Generally in the case you have described above, all badge values are only going to be acted on by the receiver, so I don't see how it can't be aware of the convention.

**Comment by canarysnort01** [comment]
> I meant that without the change, it would be possible to cap-virtualize a component by only using an endpoint and no notification object.

It's not fair to characterize this as a consequence of the change. This is a consquence of the existing kernel design -- the change does not affect the situation one way or another.

It would be fair to say that this change would prevent an interposing server from sending a message to an endpoint that the receiver would believe is a _bound notification_; but that ability is not useful for, and therefore does not prevent, doing cap-virtualization becuse the interposing server cannot duplicate the semantics of sending a bound notification.

To state it another way, the return value of the Endpoint Receive system call is conceptually a tagged union that either contains a bound notification value or an IPC message. Neither the sender of an IPC message nor the sender of a notification controls which will be delivered when the receiver next does a receive, the kernel decides. An interposing server cannot interpose on this path and decide it should send a bound notification because the sender it is interposing does not decide.

In other words, an interposing server has no mechanism to say "when this thread next does a receive send it this notification value regardless of what endpoint it waits on", other than by using the bound notification mechanism.

**Comment by canarysnort01** [comment]
@kent-mcleod Perhaps it would be helpful if I constructed a concrete counter-example showing cap-virtualization is possible with the change.

### Scenario 1: Notification virtualization

#### Objects

- Thread C: The client being cap-virtualized
- Thread S: The interposing server
- Endpoint E_notif: C believes this is a notification; S is implementing the notification interface

#### Events

1. C performs a Receive against endpoint E_notif, believing it is a notification
2. S performs a Send against E_notif, with n message conforming to the Notification interface that contains the notification word in the badge
3. C receives the message.
4. The bound notification flag **is not set** because the kernel did not deliver a bound notification. 
5. C interprets the message as a notification because that is what it was expecting in this context.

#### Discussion

- The bound notification flag did not prevent the cap-virtualization of endpoint E as a Notification.
- No additional Notification object was required to perform the virtualization.

### Scenario 2: Bound notification

#### Objects

- Thread C: The client being cap-virtualized
- Thread S: The interposing server
- Endpoint E: An endpoint
- Notification N: Bound to Thread C

#### Events

1. C performs a Receive against endpoint E, expecting to receive a message with some specific format e'
2. S signals N
3. C receives a message from the kernel formatted as a Notification
4. The bound notification flag **is set**
5. C correctly interprets the message as a Notification instead of confusing it as a message of e'

#### Discussion

- The bound notification flag prevented C from erroneously interpreting the notification message it received as if it were whatever message format it was expecting
- S must use N to send a bound notification, even without the addition of the bound notification flag, because S does not know on what object C is waiting -- it cannot implement the bound notification functionality in the way that the kernel can

### Conclusions

- A bound notification event is an exceptional case where the receiver is expecting a message with some specific format X, but is instead being delivered a message constructed by the kernel that must be interpreted as a notification.
- The kernel should indicate that it delivered a bound notification in this situation so that the receiver can correctly interpret the message as a notification instead of whatever message it was expecting
- Without a kernel-set bound notification flag the receiver **must unsafely infer it** by some means, such as having some application-defined and thus error-prone convention on what badge bits mean. While this may be an appropriate trade-off for static systems, this is a concern for large or dynamic systems.
- It is not necessary to allow senders to pretend to send a bound notification because only the kernel can send bound notifications. A bound notification is not a _property of the message itself_, but an indicator of an execeptional event by the kernel.
- The bound notification flag does not prevent cap-virtualization.

**Comment by pingerino** [comment]
@canarysnort01 so you don't want to trust the system initialiser, or dynamic system manager, to set up badges correctly such that you can distinguish between notifications and IPC messages. How then can you trust the same system manager to set up client identification via the badge? What's the difference?

**Comment by canarysnort01** [comment]
@pingerino The concern is with trusting that all applications running on the system all agree on the same convention and do not violate it even as they share endpoints and notifications amongst themselves and the system evolves dynamically.

**Comment by pingerino** [comment]
@canarysnort01 you really can't share subsystems that don't agree on badge conventions. You need something to set a sane baseline for your system. The kernel doesn't provide this.

**Comment by canarysnort01** [comment]
@pingerino I acknowledge and understand that the recommendend solution for this scenario is that we document a convention that says, system-wide, the high bit in the badge shall be used as the "this is a notification message" flag, that our system call stubs should assume this convention when it receives a message, and that any application that doesn't follow this convention does so at its own peril.

I understand and agree that the kernel cannot and should not enforce how applications use badge values.

I'm saying that using the badge value for this specific purpose is unnecessary and invites errors in a way that would be avoided if the kernel set its own flag.

**Comment by canarysnort01** [comment]
> So if you were to receive on an endpoint that you didn't control the badge allocation for, but also had a bound notification that you did control the badge allocation for, you couldn't trust that a message you were receiving was from the notification or endpoint.
> Are you able to be more specific? Generally in the case you have described above, all badge values are only going to be acted on by the receiver, so I don't see how it can't be aware of the convention.

@kent-mcleod I agree that it should be aware of the convention; I think the convention should be system-wide, and all applications should follow the convention.

This is an example of how an application could run into trouble even if it followed the convention but some other application sharing a capability with it did not.

I am trying to say that if there were a kernel flag, this convention would not be necessary, the system call stub could indicate to the caller whether it received a bound notification or a message without interpreting (and being able to misinterpret) the badge value, and so it would eliminate a possible error condition by construction.

I understand that may not be enough motiviation for you to change the design, but I hope I have at least been able to effectively communicate that this change is feasible and that it has value.

**Comment by kent-mcleod** [comment]
> It's not fair to characterize this as a consequence of the change. This is a consquence of the existing kernel design -- the change does not affect the situation one way or another.

Sorry, I should have been more clear about what my baseline was. In this case the baseline was assuming that message info register was being zeroed by notification delivery and there wasn't a way for a receiver to tell the source of the message other than the badge value.

I do agree that there is a reasonable argument that the kernel provide a way to distinguish badges that come from a bound notification and waiting on an endpoint and not rely on the badge space being partitioned internally.  However, as @pingerino points out, this would be pushing what is currently supported as a user policy down to a kernel policy and doesn't follow the typical anti-policy kernel design goals.

As @canarysnort01 points out, the current implementation supports this by not overwriting the message info register.  In my opinion this eventually should be resolved by choosing to either explicitly add a way to tell the messages apart, or to remove all abilities to tell the messages apart. I don't think that the kernel design has a concept of a receiving message coming from more than one source outside of the badge though. At this point it would be helpful to look into what changes to the specs each option would require. 

Responding to @canarysnort01 :
> A bound notification event is an exceptional case where the receiver is expecting a message with some specific format X, but is instead being delivered a message constructed by the kernel that must be interpreted as a notification.
> It is not necessary to allow senders to pretend to send a bound notification because only the kernel can send bound notifications. A bound notification is not a property of the message itself, but an indicator of an execeptional event by the kernel.

I don't agree that bound notifications are more execeptional than regular notifications.  Bound notification's aren't expected to be used only in exceptional cases.  They are also necessary to allow a single thread to receive on both an endpoint and notification at the same time.  This is necessary to implement single-threaded event-based driver components that need to handle both async events (interrupts) and sync events (calls from clients of the driver).

> The bound notification flag does not prevent cap-virtualization.
I agree.

**Comment by canarysnort01** [comment]
> Bound notification's aren't expected to be used only in exceptional cases.

@kent-mcleod You're right, I shouldn't have described them like that. I should have said that they are a special case in the receive path where there is only one valid interpretation of the message, as a notification.

**Comment by canarysnort01** [comment]
>  However, as @pingerino points out, this would be pushing what is currently supported as a user policy down to a kernel policy and doesn't follow the typical anti-policy kernel design goals.

I understand your concern. However, I don't see that there is any meaningful user policy decision in how to deal with bound notifications, other than that there exists a way to detect them.

**Comment by WildCryptoFox** [comment]
Without bound notifications, users would implement the behavior using additional threads and could distinguish the notification and endpoint. A naive implementation could use shared memory or send multiple messages to add the additional framing to preserve the union tag.

Bound notifications, as any abstraction, should only reduce the overheads and should not omit the union tag.

Without the tag, a dynamic system cannot trust delegated endpoints while using bound notifications due to the potential of confused deputies.

**Comment by kent-mcleod** [comment]
I checked with @lsf37, who indicated that proof updates for some of the various changes we've discussed shouldn't be large.  So maybe the next step is to propose this via an RFC? What do you think @pingerino, @canarysnort01?

**Comment by gernotheiser** [comment]
@WildCryptoFox would you please explain the confused deputy problem here?

**Comment by canarysnort01** [comment]
@kent-mcleod I could write an RFC to add a kernel indicator for a bound notification. Do you want to continue this discussion as a Pre-RFC on the discourse?

**Comment by WildCryptoFox** [comment]
@gernotheiser If thread A receives an endpoint from an untrusted thread B, it doesn't control the badges it may receive. Thus it cannot rely on badge space partitioning to distinguish the endpoint from a bound notification; which means the endpoint can impersonate the bound notification.

You could argue we shouldn't receive on untrusted endpoints because they might just never send messages, but we can use non-blocking receives or spawn blocking receive threads to work around the potential DoS. But if the latter cannot distinguish endpoints and bound notifications, then we would need another communication channel to certify the signal to detect impersonation.

**Comment by kent-mcleod** [comment]
> Do you want to continue this discussion as a Pre-RFC on the discourse?

Ok.

**Comment by gernotheiser** [comment]
@WildCryptoFox I'm not convinced that this scenario is reasonable enough to justify modifying kernel functionality. In particular, I'm not convinced there is a reasonable use case that requires receiving on endpoints received from an untrusted source.
